### PR TITLE
[Xtensa] Add basic support for inline asm constraints.

### PR DESCRIPTION
**Pull request by andreisfr**

### CODE REVIEW

File: llvm/test/CodeGen/Xtensa/inline-asm.ll [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+define i32 @constraint_r(i32 %a) {
+; XTENSA-LABEL: constraint_r:
+; XTENSA:         l32r a8, .LCPI0_0
+; XTENSA-NEXT:    l32i a8, a8, 0
+; XTENSA-NEXT:    #APP
+; XTENSA-NEXT:    add a2, a2, a8
+; XTENSA-NEXT:    #NO_APP
+; XTENSA-NEXT:    ret
+  %1 = load i32, i32* @gi
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/inline-asm.ll:15]: Use opaque pointers

File: llvm/test/CodeGen/Xtensa/inline-asm.ll [hunk 2]
```diff
@@ -0,0 +1,29 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 2
+; RUN: llc -mtriple=xtensa -verify-machineinstrs < %s \
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/inline-asm.ll:2]: Don't need -verify-machineinstrs

File: llvm/test/CodeGen/Xtensa/inline-asm-mem-constraint.ll [hunk 3]
```diff
@@ -0,0 +1,23 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=xtensa --verify-machineinstrs < %s | FileCheck %s --check-prefix=LA32
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/inline-asm-mem-constraint.ll:2]: Don't need -verify-machineinstrs

File: llvm/test/CodeGen/Xtensa/inline-asm.ll [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+define i32 @constraint_i(i32 %a) {
+; XTENSA-LABEL: constraint_i:
+; XTENSA:         #APP
+; XTENSA-NEXT:    addi a2, a2, 113
+; XTENSA-NEXT:    #NO_APP
+; XTENSA-NEXT:    ret
+  %1 = load i32, i32* @gi
+  %2 = tail call i32 asm "addi $0, $1, $2", "=r,r,i"(i32 %a, i32 113)
+  ret i32 %2
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/inline-asm.ll:29]: Also test some cases with physical registers, and different types?

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 5]
```diff
@@ -142,6 +142,74 @@ bool XtensaTargetLowering::isOffsetFoldingLegal(
...
+}
+
+TargetLowering::ConstraintWeight
+XtensaTargetLowering::getSingleConstraintMatchWeight(
+    AsmOperandInfo &Info, const char *Constraint) const {
+  ConstraintWeight Weight = CW_Invalid;
+  Value *CallOperandVal = Info.CallOperandVal;
+  // If we don't have a value, we can't do a match,
+  // but allow it at the lowest weight.
+  if (CallOperandVal == NULL)
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:168]: !CallOperandVal

File: llvm/test/CodeGen/Xtensa/inline-asm.ll [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+; XTENSA-NEXT:    l32i a8, a8, 0
+; XTENSA-NEXT:    #APP
+; XTENSA-NEXT:    add a2, a2, a8
+; XTENSA-NEXT:    #NO_APP
+; XTENSA-NEXT:    ret
+  %1 = load i32, i32* @gi
+  %2 = tail call i32 asm "add $0, $1, $2", "=r,r,r"(i32 %a, i32 %1)
+  ret i32 %2
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/Xtensa/inline-asm.ll:19]: Also test an unrecognized constraint code for an error

File: llvm/test/CodeGen/Xtensa/inline-asm.ll (hunk 4 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/inline-asm.ll:29]: Thank you for comments. I added more tests.

File: llvm/test/CodeGen/Xtensa/inline-asm.ll (hunk 1 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/inline-asm.ll]: Fixed

File: llvm/test/CodeGen/Xtensa/inline-asm.ll (hunk 2 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/inline-asm.ll]: Fixed

File: llvm/test/CodeGen/Xtensa/inline-asm-mem-constraint.ll (hunk 3 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/inline-asm-mem-constraint.ll]: Fxied

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 5 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fxied

File: llvm/test/CodeGen/Xtensa/inline-asm.ll (hunk 6 above)
Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/inline-asm.ll:19]: Added test

**Review by arsenm — APPROVED** [review]

**Pull request merged**